File: frontend/src/app/models/reservation.ts
export type ReservationStatus = 'PENDING' | 'CONFIRMED' | 'CANCELLED' | 'EXPIRED' | 'FAILED';
export type PaymentStatus = 'PENDING' | 'PAID' | 'REFUNDED' | 'FAILED';

export interface AuditEntry {
  action: string;
  performedBy: number;
  timestamp: string;
  details: string;
}

export interface Reservation {
  id: string;
  userId: number;
  eventId: number;
  ticketId?: number | null;
  numberOfPlaces: number;
  totalPrice: number;
  status: ReservationStatus;
  paymentStatus: PaymentStatus;
  reservationDate: string;
  confirmationDate: string;
  cancellationDate: string;
  cancellationReason: string;
  auditLog: AuditEntry[];
  createdAt: string;
  updatedAt: string;
  username: string;
  eventTitle: string;
  eventDate: string;
}

export interface ReservationRequest {
  userId: number;
  eventId: number;
  numberOfPlaces: number;
}

export interface CancelRequest {
  reason: string;
  cancelledBy: number;
}

export interface AvailabilityResponse {
  eventId: number;
  totalPlaces: number;
  reservedPlaces: number;
  availablePlaces: number;
}

/** POST /api/saga/reserve — orchestrated distributed reservation */
export interface SagaReserveRequest {
  userId: number;
  eventId: number;
  places: number;
  idempotencyKey?: string;
  nomClient?: string;
  emailClient?: string;
}

export interface SagaReserveResponse {
  sagaId: string;
  status: string;
  reservationId?: string;
  ticketId?: number;
  message?: string;
}

File: microservices/reservation-service/src/main/java/com/example/reservationservice/document/Reservation.java
package com.example.reservationservice.document;

import lombok.*;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Document(collection = "reservations")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Reservation {

    @Id
    private String id;

    private Long userId;
    private Long eventId;
    private Long ticketId;

    private int numberOfPlaces;
    private Double totalPrice;

    @Builder.Default
    private ReservationStatus status = ReservationStatus.PENDING;

    @Builder.Default
    private PaymentStatus paymentStatus = PaymentStatus.PENDING;

    private LocalDateTime reservationDate;
    private LocalDateTime confirmationDate;
    private LocalDateTime cancellationDate;
    private String cancellationReason;

    @Builder.Default
    private List<AuditEntry> auditLog = new ArrayList<>();

    @CreatedDate
    private LocalDateTime createdAt;

    @LastModifiedDate
    private LocalDateTime updatedAt;

    public void addAuditEntry(String action, Long performedBy, String details) {
        if (auditLog == null) auditLog = new ArrayList<>();
        auditLog.add(AuditEntry.builder()
                .action(action)
                .performedBy(performedBy)
                .timestamp(LocalDateTime.now())
                .details(details)
                .build());
    }
}
